add  system.distributed_ddl_queue table

bharatnc opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Adds a new table called `system.distributed_ddl_queue` that displays the queries in the DDL worker queue.

Detailed description / Documentation draft:

This PR adds a new table called the `system.distributed_ddl_queue` that lists all the queries that are currently in the DDL worker queue.

To accomplish this, the zookeeper path for `distributed_ddl.path` (default is `/clickhouse/task_queue/ddl/`) is polled for
all the queries and for each query, subpaths `/active` and `/finished` are queried to get the list of nodes that are present under the `active` and `finished` zookeeper paths. The data is then populated into the table as follows:

Querying the `system.distributed_ddl_queue` table from one of the shards:

```sql
SELECT *
FROM system.distributed_ddl_queue
WHERE cluster = 'test_cluster'
LIMIT 1
FORMAT Vertical

Query id: a1c53d35-5a84-4e3d-8176-13b1f4791141

Row 1:
──────
entry:             query-0000000000
host_name:         clickhouse01
host_address:      172.23.0.11
port:              9000
status:            Finished
cluster:           test_cluster
query:             CREATE DATABASE test_db UUID '4a82697e-c85e-4e5b-a01e-a36f2a758456' ON CLUSTER test_cluster
initiator:         clickhouse01:9000
query_start_time:  2020-12-30 13:07:51
query_finish_time: 2020-12-30 13:07:51
query_duration_ms: 6
exception_code:    ZOK

1 rows in set. Elapsed: 0.038 sec. 
```

relates to #17082

bharatnc committed:
DDLWorkerQueueTable - add intial scaffolding

bharatnc committed:
DDLWorkerQueueTable - initial attempt to fetch entries from zk

bharatnc committed:
DLLWorkerQueueTable - remove unimplemented function

bharatnc committed:
DDLWorkerQueueTable - populate active and finished columns

bharatnc committed:
DDlWorkerQueueTable - add integration test

bharatnc committed:
DDLWorkerQueueTable - load distributed_ddl.path from server config

bharatnc committed:
DDLWorkerQueueTable - make loop vars a const ref

bharatnc committed:
DDLWorkerQueueTable - fix integration test

bharatnc commented:
@tavplubix , @alesapin  this is my current implementation of the `ddl_worker_queue` table. It's now ready for review.

The table structure is as follows:

```text
:) describe system.ddl_worker_queue;

DESCRIBE TABLE system.ddl_worker_queue

Query id: 5c760afb-f63c-4ee9-b359-e4ee38c663f9

┌─name─────┬─type──────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ name     │ String        │              │                    │         │                  │                │
│ active   │ Array(String) │              │                    │         │                  │                │
│ finished │ Array(String) │              │                    │         │                  │                │
└──────────┴───────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

3 rows in set. Elapsed: 0.004 sec. 
```

Also I rely on zookeeper for fetching a list of  queries and the entries under active and finished paths.  I'm pretty sure that there might be a better way to accomplish this end result and also the table can include more columns probably or even the structure perhaps might not be the one that you all would have visualized. Do let me know in your reviews.

azat commented on src/Storages/System/StorageSystemDDLWorkerQueue.cpp:
```diff
@@ -0,0 +1,75 @@
+{
+    zkutil::ZooKeeperPtr zookeeper = context.getZooKeeper();
+
+    String ddl_zookeeper_path = config.getString("distributed_ddl.path", "/clickhouse/task_queue/ddl/");
```
Should use `distributed_ddl.path` config directive

bharatnc replied:
@azat not sure if I understand you completely. Are you implying if I need to a check similar to:`if ( config().has("distributed_ddl"))` before actually doing a `config.getString(...)` ?

azat replied:
Egh, sorry, by some reason I missed that out already did this

azat reviewed (commented):
>also the table can include more columns probably

How about adding the following columns?
- query itself (from `query-X` znode)
- query_duration_ms (`mtime` for the parent znode (`query-X`) and `query-X/finished/{node}`)
- exception_code (`finished/{node}` znode)

This way it can be used to track errors and is query executed eventually (since right now if the timeout will be reached you will not know the status of the query in the client)

Also since znodes in zookeeper periodically cleaned up, looks like adding `system.ddl_worker_queue_log` seems useful (but can be done as a separate step I guess), but it should not be enabled by default of course (since the content is the same for all cluster)

And by the way it worth mention this new table in `system-tables.md`

tavplubix commented:
Maybe unfold arrays in table structure? I mean something like:
```
┌─entry────────────┬─host───────┬─status──────┬─some_column─┐
│ query-0000000042 │ node1:9000 │ active      │ ...         │
│ query-0000000042 │ node2:9000 │ finished    │ ...         │
│ query-0000000042 │ node3:9000 │ unknown     │ ...         │
│ ...              │ ...        │ ...         │ ...         │
└──────────────────┴────────────┴─────────────┴─────────────┘
```

And I suggest the following columns:
 - `entry String` (or `entry_name`) instead of `name`
 - `cluster String` - cluster name, we can extract it from `query`, but it's easier to use `where cluster='smth'` in `select` 
 - `host String` (or maybe two separate columns with host and port) - name of host, which should execute the query. We can obtain it from hosts list in DDL entry (`query-xxxxxxxxxx` node data). However, hosts list in DDL entry contains IP addresses rather than hostnames, so I'm not sure if it will be convenient to use. We also can try to resolve it back to hostnames (see `DDLWorker::parseQueryAndResolveHost(...)`), but it's too complicated. Anyway, `DDLQueryStatusInputStream` relies on this hosts list when it waiting for query to be executed on all hosts, so we should take it into account and show all involved hosts in `system.ddl_worker_queue`, even if some of them have not started to execute query yet (and have not created corresponding `active` node yet). 
 - `status Enum('unknown' = 1, 'active' = 2, 'finished' = 3, 'error' = 4, ...)` - status of query execution on each host, `unknown` is for hosts which have not created either `active` or `finished` node in zk, `error` is for `finished` nodes with non-zero exception code.
 - `query`, `query_duration_ms` and `exception_code` will be useful too
 - Maybe some additional columns with name of replica which executes query and `tries_to_execute` count (if query should be executed on leader, see `DDLWorker::tryExecuteQueryOnLeaderReplica(...)`)


As for `system.ddl_worker_queue_log`, now all such queries are logged into `system.query_log` with special comment, so it's possible to get query execution status with `select ... from clusterAllReplicas('cluster', 'system.query_log') where query like '/* ddl_entry=query-xxxxxxxxxx */%'`. But it's a bit different thing and I agree that `system.ddl_worker_queue_log` maybe also useful in some cases.

Another consideration: `DDLWorker` is a name of class in code, so name `ddl_worker_queue` is clear for developers, but not for users. Maybe `distributed_ddl_queue` or `on_cluster_ddl_queue` is more suitable? 

@azat, @alesapin, do you have any comments?

tavplubix commented on tests/integration/test_system_ddl_worker_queue/test.py:
```diff
@@ -0,0 +1,63 @@
+        test_cluster.instances['ch{}'.format(i + 1)].query("INSERT INTO merge (i) VALUES ({})({})".format(k, k + 1))
+    time.sleep(2)
+    test_cluster.ddl_check_query(instance, "ALTER TABLE merge ON CLUSTER '{cluster}' DETACH PARTITION 197002")
+    time.sleep(2)
+    # query the ddl_worker_queue table to ensure that the columns are populated as expected
+    assert "ok\n" in instance.query(
+        "SELECT If((SELECT count(finished) FROM system.ddl_worker_queue WHERE name='query-0000000000') > 0, 'ok', 'fail')");
+    assert "ok\n" in instance.query(
+        "SELECT If((SELECT count(active) FROM system.ddl_worker_queue WHERE name='query-0000000000') >= 0, 'ok', 'fail')")
```
Tests with `sleep` are usually flacky, better to use something like:
```
from helpers.test_tools import assert_eq_with_retry
...
assert_eq_with_retry(instance, "SELECT ...", "ok")
```

tavplubix commented on tests/integration/test_system_ddl_worker_queue/test.py:
```diff
@@ -0,0 +1,63 @@
+    for i in range(0, 4, 2):
+        k = (i / 2) * 2
+        test_cluster.instances['ch{}'.format(i + 1)].query("INSERT INTO merge (i) VALUES ({})({})".format(k, k + 1))
+    time.sleep(2)
```
```
for instance in instances:
    instance.query("SYSTEM SYNC REPLICA merge")
```

azat commented:
>host String (or maybe two separate columns with host and port) - name of host, which should execute the query

`system.clusters` has `host_address`/`host_name` and `port`, so better to use the same naming.

>Maybe unfold arrays in table structure? I mean something like:

Completely agree with everything and unfolding arrays in particular.

tavplubix commented on src/Storages/System/StorageSystemDDLWorkerQueue.cpp:
```diff
@@ -0,0 +1,75 @@
+       */
+        size_t col_num = 0;
+        ddl_query_path = ddl_zookeeper_path + "/" + q;
+        zkutil::Strings active_nodes = zookeeper->getChildren(ddl_query_path + "/active");
```
`select ... from system.ddl_worker_queue` may fail with `ZNONODE` in the following cases:
 - Initiator of `ON CLUSTER` query has lost connection [here](https://github.com/ClickHouse/ClickHouse/blob/2e9f10ab3ec8bc99b6b9596e7cb0b9660195fefb/src/Interpreters/DDLWorker.cpp#L1096) and all hosts are busy with older queue entries, so status dirs are not [created](https://github.com/ClickHouse/ClickHouse/blob/2e9f10ab3ec8bc99b6b9596e7cb0b9660195fefb/src/Interpreters/DDLWorker.cpp#L760) yet
 - race with `DDLWorker::runCleanupThread()`

The simplest way to fix it - use `tryGetChildren(...)` and leave some columns empty if we cannot get data from zk. 
Another way is to use [locks in zk](https://github.com/ClickHouse/ClickHouse/blob/2e9f10ab3ec8bc99b6b9596e7cb0b9660195fefb/src/Interpreters/DDLWorker.cpp#L1025), but let's try to do it without locks

azat replied:
>The simplest way to fix it - use tryGetChildren(...) and leave some columns empty if we cannot get data from zk.

But this way the data will be unreliable, no?
At least all columns should be marked as `Nullable`, and `Null` means data retrieving error, @tavplubix what do you think?

tavplubix replied:
> data will be unreliable

Not precisely, `status` will be `unknown` and and some columns (`query_duration_ms`, `exception`) will be empty. However, i'm not sure if it's possible to distinguish the first case (no one has started to execute entry) with the second (entry are currently removed). 

And of course, we should not provide incomplete data on zk errors other than `ZNONODE`.

bharatnc replied:
I have changed to use `tryGetChildren(...)` as suggested.

bharatnc commented:
Thank you for the review @azat  and @tavplubix let me take sometime to read the reviews more closer and also make these changes.

bharatnc committed:
Review - initial round of changes

bharatnc committed:
Add values column and fix tests

bharatnc committed:
Use tryGetChildren() for Zookeeper

bharatnc committed:
Add query_start|finish_time, query_dureation_ms

bharatnc committed:
add exception_status column

bharatnc committed:
few more changes and simplification to integration test

bharatnc committed:
Docs - distributed_ddl_queue

bharatnc commented:
@tavplubix  and @azat 

I have implemented the following as per your review:

- [X] entry String (or entry_name) instead of name
- [X] cluster String - cluster name, we can extract it from query, but it's easier to use where cluster='smth' in select
- [X] host String (or maybe two separate columns with host and port) - name of host, which should execute the query .....
- [X]  status Enum('unknown' = 1, 'active' = 2, 'finished' = 3, 'error' = 4, ...) - status of query execution on each host, .....
- [X] query, query_duration_ms and exception_code will be useful too
- [X] Maybe some additional columns with name of replica which executes query and tries_to_execute count (if query should be executed on leader, see DDLWorker::tryExecuteQueryOnLeaderReplica(...))
- [X] - replace `getChildren()` with `tryGetChildren()` 
- [X] - using the table name as `system.distributed_ddl_queue` - Let me know if you want me to change the file names to match as well. But I think DDLWorkerQueue sounds clearer. for developers as pointed out.

Additional points:
- Current query output looks like the following:
```sql

Row 1:
──────
entry:             query-0000000000
host_name:         clickhouse01
host_address:      172.23.0.11
port:              9000
status:            finished
cluster:           test_cluster
values:            version: 1
query: CREATE DATABASE test_db UUID '40ac7692-70d3-48a9-bc29-4ade18957f59' ON CLUSTER test_cluster
hosts: ['clickhouse01:9000','clickhouse02:9000','clickhouse03:9000','clickhouse04:9000']
initiator: clickhouse01:9000
```
- I've written the checks in such a way that if any one of the columns are unavailable perhaps, due to exception, then the columns are populated with best effort.

Concerns:
- I'm worried about the complexity of the code that I've written. The current flow looks like:
  - I'm now using `WHERE cluster='cluster'` in the query.
  - For each `cluster` I'm querying all replicas.
  - For all the queries, I'm populating all the relevant fields using the replicas and also iterating through a list of queries to 
    determine the status of the queries.

Let me know if the implementation looks better now (but it's doing more now). Thank you!

bharatnc commented on tests/integration/test_system_ddl_worker_queue/test.py:
```diff
@@ -0,0 +1,48 @@
+
+    node1.query("ALTER TABLE testdb.test_table ON CLUSTER test_cluster ADD COLUMN somecolumn UInt8 AFTER val",
+                settings={"replication_alter_partitions_sync": "2"})
+    for node in nodes:
```
Tests could be better, but I can follow up with improvements in follow up PRs.

bharatnc committed:
review fixes

tavplubix commented on src/Storages/System/StorageSystemDDLWorkerQueue.h:
```diff
@@ -0,0 +1,33 @@
+class Context;
+
+
+/** System table "ddl_worker_queue" with list of queries that are currently in the DDL worker queue.
```
```suggestion
/** System table "distributed_ddl_queue" with list of queries that are currently in the DDL worker queue.
```

tavplubix commented on src/Storages/System/StorageSystemDDLWorkerQueue.cpp:
```diff
@@ -0,0 +1,289 @@
+void StorageSystemDDLWorkerQueue::fillData(MutableColumns & res_columns, const Context & context, const SelectQueryInfo & query_info) const
+{
+    String cluster_name = extractPath(query_info.query, context);
+    if (cluster_name.empty())
+        throw Exception(
+            "SELECT from system.ddl_worker_queue table must contain condition like cluster = 'cluster' in WHERE clause.",
```
Why is it necessary? Maybe show queries for all clusters? If query contains WHERE clause with condition on cluster name, all irrelevant rows will be filtered later in query pipeline.

bharatnc replied:
Sure, I can change.

bharatnc replied:
Showing all clusters and the necessary clusters can be filtered later in the pipeline as suggested.

tavplubix commented on src/Storages/System/StorageSystemDDLWorkerQueue.cpp:
```diff
@@ -0,0 +1,289 @@
+}
+
+
+static bool extractPathImpl(const IAST & elem, String & res, const Context & context)
```
Looks like copy-paste from `StorageSystemZooKeeper.cpp`

bharatnc replied:
Yes, it is. I was going to leave a TODO to extract this function common to both files  (they currently use)in a follow up PR. Those functions are private to the `StorageSystemZookeeper.cpp` and also use a different ident `path` instead of `cluster`. I can make those changes in the same PR or can follow up.

bharatnc replied:
ended up removing this code as it's no longer needed.

tavplubix commented on src/Storages/System/StorageSystemDDLWorkerQueue.cpp:
```diff
@@ -0,0 +1,289 @@
+       [query-0000000000, query-0000000001, query-0000000002, query-0000000003, query-0000000004]
+    */
+
+    // if there is error querying the root path of the ddl directory throw exception and stop
```
But now it returns 0 rows instead of throwing exception if we cannot get list of entries in queue

bharatnc replied:
Yes, this comment is not accurate has to be removed. Initially I had thought about throwing an exception if we aren't able to query node information anywhere in the code flow. But if we cannot retrieve the list of entries for a single node then we won't be able to fill the `distributed_ddl_queue` for all other nodes as well.
I will remove the comment and instead leave the rows empty on failure.

tavplubix commented on src/Storages/System/StorageSystemDDLWorkerQueue.cpp:
```diff
@@ -0,0 +1,289 @@
+                }
+
+                //  This is the original cluster_name from the query (WHERE cluster='cluster_name')
+                res_columns[i++]->insert(cluster_name);
```
We should obtain `cluster_name` from DDL query (the query that is written into `/clickhouse/task_queue/ddl/query-0000000000` node), not from `SELECT ... FROM system.distributed_ddl_queue`

bharatnc replied:
`cluster_name` now extracted from DDL query (referred from DDLWorker.cpp)

tavplubix commented on docs/en/operations/system-tables/distributed_ddl_queue.md:
```diff
@@ -0,0 +1,72 @@
+port:              9000
+status:            finished
+cluster:           test_cluster
+value:            version: 1
```
Why not to parse the value using `DDLLogEntry::parse(...)` (see DDLWorker.cpp) and not to retrieve query from it? Also I'm not sure if we should print hosts list here, because we have `host_name`, `host_address` and `port` columns to print it.

bharatnc replied:
Thanks for the pointers, I can look into it. I agree that printing hosts here looks confusing.

bharatnc replied:
> Why not to parse the value using DDLLogEntry::parse(...) (see DDLWorker.cpp) and not to retrieve query from it?

I did not get this part:
What do you mean by "not to retrieve query from it" ? Do you mean use DDLLogEntry::parse(...)  to only retrieve relevant stuff like `query`, `version` and `initiator`?

@tavplubix if you could explain, that would be helpful.

tavplubix replied:
I mean "why ... not to retrieve query from it". Yes, we should parse the content of `query-xxxxxx` node using `DDLLogEntry::parse(...)` to retrieve `query`, `initiator` and `hosts`. And we should print `query`, `initiator` and `hosts` in separate columns, not in a single `value` column with plain-text content of `query-xxxxxxx`. 

As for `hosts`, we already have `host_name`, `host_address` and `port` columns. Values in these columns must be consistent with the hosts list parsed from `query-xxxxxxx`

bharatnc replied:
Thanks for clarifying I have made these changes.

bharatnc committed:
DDLLogEntry.parse() to populate query and initiator

bharatnc committed:
extract cluster_name from DDLQuery

bharatnc commented:
@tavplubix  pushed a second round of changes since your last review and have updated the PR description to contain the latest o/p. Let me know if it looks good to you. I hope that I've addressed most of the concerns. 

One note about exception handling, if anywhere a zk exception occurs, the exception code is updated and we leave the column empty instead of throwing that exception. This is so that we don't fail on all results one of the zk query fails / timeouts.

bharatnc committed:
Docs - update to reflect code changes

bharatnc committed:
Fixes to code, build and style checks

azat commented on src/Storages/System/StorageSystemDDLWorkerQueue.cpp:
```diff
@@ -0,0 +1,235 @@
+std::vector<std::pair<String, Int8>> getStatusEnumsAndValues()
+{
+    return std::vector<std::pair<String, Int8>>{
+        {"active", static_cast<Int8>(Status::active)},
+        {"finished", static_cast<Int8>(Status::finished)},
+        {"unknown", static_cast<Int8>(Status::unknown)},
+        {"errored", static_cast<Int8>(Status::errored)},
```
Looks like enums in system tables uses CamelCase syntax (query_log/text_log)

bharatnc replied:
sure, changed to CamelCase.

azat commented on src/Storages/System/StorageSystemDDLWorkerQueue.cpp:
```diff
@@ -0,0 +1,235 @@
+                    zkutil::Strings active_nodes;
+                    zkutil::Strings finished_nodes;
+
+                    // on error just skip and continue.
```
Looks like this is not true anymore (f8af76ba362c0710c473438c01bdc159fd6efb63)

bharatnc commented:
@tavplubix @azat just following up - let me know if you have any other inputs, I have implemented the changes based on your latest reviews.

alexey-milovidov commented:
@tavplubix is on vacation until Jan 11.

azat commented on src/Storages/System/StorageSystemDDLWorkerQueue.cpp:
```diff
@@ -0,0 +1,232 @@
+                    }
+
+                    Coordination::GetResponse res;
+                    if (!futures.empty())
```
BTW is this possible? You have reserve() before

bharatnc replied:
I think it's not necessary - removed it.

azat commented on src/Interpreters/DDLWorker.h:
```diff
@@ -1,12 +1,15 @@
 #include <Interpreters/Cluster.h>
 #include <Interpreters/Context.h>
-#include <DataStreams/BlockIO.h>
+#include <Storages/IStorage.h>
```
```suggestion
#include <Storages/IStorage_fwd.h>
```

It does not use anything from the IStorage, it just need the declaration

azat reviewed (approved):
LGTM
